qa/tasks/cephfs/test_nfs.py

```diff
@@ -876,3 +876,75 @@ def test_nfs_export_creation_at_symlink(self):
+        """
+        Test that on export update/creation failure while using
+        ceph nfs export apply <nfs_cluster> -i <json/conf>, the failed
+        exports export_id are visible in the JSON response to CLI.
+        """
+        mnt_pt = self._sys_cmd(['mktemp', '-d']).decode().strip()
+        self._create_cluster_with_fs(self.fs_name, mnt_pt)
+        self.ctx.cluster.run(args=['mkdir', f'{mnt_pt}/testdir1'])
+        self.ctx.cluster.run(args=['mkdir', f'{mnt_pt}/testdir2'])
+        self._create_export(export_id='1',
+                            extra_cmd=['--pseudo-path', self.pseudo_path])
+
+        ret = self.ctx.cluster.run(args=['ceph', 'nfs', 'export', 'apply',
+                                         self.cluster_id, '-i', '-'],
+                                   check_status=False,
+                                   stdin=json.dumps([
+                                    {
+                                        "export_id": 11,
+                                        "path": "/testdir1",
+                                        "pseudo": self.pseudo_path,
+                                        "squash": "none",
+                                        "access_type": "rw",
+                                        "protocols": [4],
+                                        "fsal": {
+                                            "name": "CEPH",
+                                            "fs_name": self.fs_name
+                                        }
+                                    },
+                                    {
+                                        "path": "/testdir2",
+                                        "pseudo": f"/{self.pseudo_path}1",
+                                        "squash": "none",
+                                        "access_type": "rw",
+                                        "protocols": [4],
+                                        "fsal": {
+                                            "name": "CEPH",
+                                            "fs_name": "foo"
+                                        }
+                                    }]),
+                                   stdout=StringIO(), stderr=StringIO())
+
+        try:
+            err_info = ret[0].stdout
+            if err_info:
+                update_details = json.loads(err_info.getvalue())
+
+                self.assertEqual(update_details[0]["export_id"], 1)
+                self.assertIn("Export ID changed, Cannot update export",
+                              update_details[0]["msg"])
+
+                # the second export block provided was a new export creation
+                # request, so even though the export_id wasn't mentioned, 2
+                # is the next available int value for exports since the only existing export has
+                # export id: 1, therefore it must report 2 in the dict.
+                self.assertEqual(update_details[1]["export_id"], 2)
```

Comment:
How is this useful to the user who is passing a export block to create a new export, which fails, but the export_id in present in the output json.

Reply:
it might not be a big deal if there is a single block or a few blocks but if there are extensive exports then it will helps figure out which export block creation/update failed

Reply:
Sure, that's fine for existing exports, but for exports which are new and are failing creation, the export_id isn't of any use, right?

Reply:
... and following up on this: On the next call if the export creation is successful, what export_id is assigned?

Reply:
if it's a new export creation request, a call is made:
https://github.com/ceph/ceph/blob/d759fe4103174d412cc1ba317390383b299a7d35/src/pybind/mgr/nfs/export.py#L780-L784

if you see, before processing anything in `create_export_from_dict()`, it provides an ID to the export, if everything goes fine, the code block below is executed which saves the new export to rados else the exception is raised and thus nothing is committed anywhere:
https://github.com/ceph/ceph/blob/d759fe4103174d412cc1ba317390383b299a7d35/src/pybind/mgr/nfs/export.py#L786-L789

this means when the export creation failed, we did have the export_id of the new export that was to be created, so by having the export_id in the JSON output (of the export that failed to create) we mean to say "hey you wanted to create an export but didn't provide an export id, that's fine we gave it X which was the next available integer but due to something in the export block, it(X) could not be created".

Reply:
> ... and following up on this: On the next call if the export creation is successful, what export_id is assigned?

the code that finds the next available ID is:
https://github.com/ceph/ceph/blob/d759fe4103174d412cc1ba317390383b299a7d35/src/pybind/mgr/nfs/export.py#L282-L290

i.e. it checks the nfs cluster; goes through all the exports and starting from 1 it returns the next available integer that can be used. so in case of export creation failure, nothing is committed in the cluster and thus next time it will start from `(last export_id in cluster)+1`

Reply:
and on top of it you have this line:https://github.com/ceph/ceph/blob/d759fe4103174d412cc1ba317390383b299a7d35/src/pybind/mgr/nfs/export.py#L637

so now let's say have a export block like and you already have an export in cluster with export_id 1: 
```
  {
    "path": "dir1",
    "cluster_id": "my-nfs",
    "pseudo": "/ceph",
    "access_type": "RW",
    "squash": "none",
    "security_label": true,
    "protocols": [
      4
    ],
    "transports": [
      "TCP"
    ],
    "fsal": {
      "name": "CEPH",
      "user_id": "foo",  -----> invalid
      "fs_name": "myfs"
    },
    "clients": []
  }
```


, the output json would be:
```
[
  {
    "msg": "Failed to apply export: export FSAL user_id must be 'nfs.test.2'",
    "state": "error"
  }
]
```

the 2 in suffix is the export_id.

Reply:
the point is anyone who is using this should be knowing/must know about this behaviour that we provide export_id to the export(if not provided) depending on the exports in the cluster. This is also a good time to add this in the NFS docs as a note

Reply:
Yeh, I understand the code part. My question was more of related to the export_id returned in the JSON on failure. I understand that the export_id isn't committed. Maybe returning the export_id is fine since its unused but I hope users do not get confused with this behaviour.

Reply:
removing export_id for export creation failure will deteriorate the consistency of the JSON output which john feels is a bummer. So if we want to move ahead with having export_id then we will have to make some trade-offs or maybe we can make the export_id mandatory in the conf/json file?

If not export_id then we can have pseudo instead, it is a mandatory field so the user can relate and figure out which export block it was in the conf/json file.

Reply:
> removing export_id for export creation failure will deteriorate the consistency of the JSON output which john feels is a bummer. So if we want to move ahead with having export_id then we will have to make some trade-offs or maybe we can make the export_id mandatory in the conf/json file?
> 
> If not export_id then we can have pseudo instead, it is a mandatory field so the user can relate and figure out which export block it was in the conf/json file.

You mean pseudo_path, isn't it? That sounds  fine to me.

Reply:
@dparmar18 Let's have this change since it does not confuses users.

Reply:
> > removing export_id for export creation failure will deteriorate the consistency of the JSON output which john feels is a bummer. So if we want to move ahead with having export_id then we will have to make some trade-offs or maybe we can make the export_id mandatory in the conf/json file?
> > If not export_id then we can have pseudo instead, it is a mandatory field so the user can relate and figure out which export block it was in the conf/json file.
> 
> You mean pseudo_path, isn't it? That sounds fine to me.

yes pseudo is pseudo path, we don't have anything like pseudo_path